server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java

```diff
@@ -43,10 +52,57 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext) {
      * @return a new {@link MapperMergeContext} with this context as its parent
      */
     public MapperMergeContext createChildContext(String name) {
-        return new MapperMergeContext(mapperBuilderContext.createChildContext(name));
+        return createChildContext(mapperBuilderContext.createChildContext(name));
+    }
+
+    /**
+     * Creates a new {@link MapperMergeContext} with a given {@link MapperBuilderContext}
+     * @param childContext the child {@link MapperBuilderContext}
+     * @return a new {@link MapperMergeContext}, wrapping the provided {@link MapperBuilderContext}
+     */
+    public MapperMergeContext createChildContext(MapperBuilderContext childContext) {
+        return new MapperMergeContext(childContext, remainingFieldsUntilLimit);
     }
 
     MapperBuilderContext getMapperBuilderContext() {
         return mapperBuilderContext;
     }
+
+    void removeRuntimeField(Map<String, RuntimeField> runtimeFields, String name) {
+        if (runtimeFields.containsKey(name)) {
+            runtimeFields.remove(name);
+            if (remainingFieldsUntilLimit.get() != Long.MAX_VALUE) {
+                remainingFieldsUntilLimit.incrementAndGet();
+            }
+        }
+    }
+
+    void addRuntimeFieldIfPossible(Map<String, RuntimeField> runtimeFields, RuntimeField runtimeField) {
+        if (runtimeFields.containsKey(runtimeField.name())) {
+            runtimeFields.put(runtimeField.name(), runtimeField);
+        } else if (canAddField(1)) {
+            remainingFieldsUntilLimit.decrementAndGet();
+            runtimeFields.put(runtimeField.name(), runtimeField);
+        }
+    }
+
+    boolean addFieldIfPossible(Map<String, Mapper> mappers, Mapper mapper) {
+        if (canAddField(mapper.mapperSize())) {
+            remainingFieldsUntilLimit.getAndAdd(mapper.mapperSize() * -1);
+            mappers.put(mapper.simpleName(), mapper);
+            return true;
+        }
+        return false;
+    }
+
+    void addFieldIfPossible(Mapper mapper, Runnable addField) {
```

Comment:
The reason why I suggested the boolean approach is that we have now two methods that do the same, but not quite. I think it would be nice to consolidate that to a single method if possible. After all, if we return a boolean in one method, it is already responsibility of the caller to act on the return value, but the budget check is done within the method, where it belongs.

Reply:
> The reason why I suggested the boolean approach is that we have now two methods that do the same, but not quite. I think it would be nice to consolidate that to a single method if possible.

The new `NewFieldsBudget` interface now has a `decrementIfPossible(long fieldSize)` method that does what you proposed. I've still kept the `addFieldIfPossible(M mapper, Consumer<M> addField)` method as the only public method in `MapperMergeContext` that internally calls `decrementIfPossible(long fieldSize)` and also calls the consumer if there was enough budget. I do like the encapsulation of checking the budget, decrementing the budget, and adding the Mapper to the map in a single call.

>  After all, if we return a boolean in one method, it is already responsibility of the caller to act on the return value, but the budget check is done within the method, where it belongs.

In most cases, the caller doesn't need to know whether the Consumer has been called. But in some instances, it's important. But I think it shouldn't be the responsibility of the caller to add the mapper. It seems more brittle to me.